### ISSUE TITLE
RuntimeError: aten::grid_sampler_2d_backward() is missing value for argument 'output_mask'.

### ISSUE THREAD

**Issue by zikpefu** [pytorch/pytorch#75018]
### 🐛 Describe the bug

When training the network stylegan3-t from Nvidia: https://github.com/NVlabs/stylegan3 with command: `python train.py --outdir=training-runs-anime --cfg=stylegan3-t --data=dataset/anime_dataset.zip --gpus=1 --batch=32 --gamma=0.5 --mirror=1`

I get the following error: 
```
/home/zikpefu/ece8550/final/stylegan3/training/augment.py:231: UserWarning: Specified kernel cache directory could not be created! This disables kernel caching. Specified directory is /home/zikpefu/.cache/torch/kernels. This warning will appear only once per process. (Triggered internally at  ../aten/src/ATen/native/cuda/jit_utils.cpp:860.)
  s = torch.exp2(torch.randn([batch_size], device=device) * self.scale_std)
Traceback (most recent call last):
  File "/home/zikpefu/ece8550/final/stylegan3/train.py", line 286, in <module>
    main() # pylint: disable=no-value-for-parameter
  File "/home/zikpefu/.conda/envs/pytorch/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/zikpefu/.conda/envs/pytorch/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/zikpefu/.conda/envs/pytorch/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/zikpefu/.conda/envs/pytorch/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/zikpefu/ece8550/final/stylegan3/train.py", line 281, in main
    launch_training(c=c, desc=desc, outdir=opts.outdir, dry_run=opts.dry_run)
  File "/home/zikpefu/ece8550/final/stylegan3/train.py", line 96, in launch_training
    subprocess_fn(rank=0, c=c, temp_dir=temp_dir)
  File "/home/zikpefu/ece8550/final/stylegan3/train.py", line 47, in subprocess_fn
    training_loop.training_loop(rank=rank, **c)
  File "/home/zikpefu/ece8550/final/stylegan3/training/training_loop.py", line 278, in training_loop
    loss.accumulate_gradients(phase=phase.name, real_img=real_img, real_c=real_c, gen_z=gen_z, gen_c=gen_c, gain=phase.interval, cur_nimg=cur_nimg)
  File "/home/zikpefu/ece8550/final/stylegan3/training/loss.py", line 81, in accumulate_gradients
    loss_Gmain.mean().mul(gain).backward()
  File "/home/zikpefu/.conda/envs/pytorch/lib/python3.9/site-packages/torch/_tensor.py", line 363, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/zikpefu/.conda/envs/pytorch/lib/python3.9/site-packages/torch/autograd/__init__.py", line 173, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/home/zikpefu/.conda/envs/pytorch/lib/python3.9/site-packages/torch/autograd/function.py", line 253, in apply
    return user_fn(self, *args)
  File "/home/zikpefu/ece8550/final/stylegan3/torch_utils/ops/grid_sample_gradfix.py", line 50, in backward
    grad_input, grad_grid = _GridSample2dBackward.apply(grad_output, input, grid)
  File "/home/zikpefu/ece8550/final/stylegan3/torch_utils/ops/grid_sample_gradfix.py", line 59, in forward
    grad_input, grad_grid = op(grad_output, input, grid, 0, 0, False)
RuntimeError: aten::grid_sampler_2d_backward() is missing value for argument 'output_mask'. Declaration: aten::grid_sampler_2d_backward(Tensor grad_output, Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool[2] output_mask) -> (Tensor, Tensor)
```

I will say that this did work for me before on a previous version of pytorch but I forgot which version it was (maybe 1.7?)

### Versions

```
Collecting environment information...
PyTorch version: 1.11.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 8.2.2004 (Core)  (x86_64)
GCC version: (GCC) 8.4.1 20200928 (Red Hat 8.4.1-1)
Clang version: Could not collect
CMake version: version 3.18.2
Libc version: glibc-2.28

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.18.0-193.6.3.el8_2.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 470.42.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[pip3] numpydoc==1.2
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] mypy_extensions           0.4.3            py39h06a4308_1  
[conda] numpy                     1.21.2           py39h20f2e39_0  
[conda] numpy-base                1.21.2           py39h79a1101_0  
[conda] numpydoc                  1.2                pyhd3eb1b0_0  
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     1.11.0                   pypi_0    pypi
[conda] torchaudio                0.11.0                   pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.12.0                   pypi_0    pypi
```

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @gchanan

**Comment by mruberry** [comment]
cc @to-mi @albanD

**Comment by to-mi** [comment]
My (non-expert) analysis of this: There was a backwards-incompatible change to `grid_sampler_2d_backward` in PyTorch 1.11.0 (it now takes in the grad mask to possibly avoid computing unnecessary gradient for the input) and the stylegan3 code is calling this function directly. The commits for this change are here: https://github.com/pytorch/pytorch/commits?author=to-mi&since=2021-09-30&until=2021-10-31

Some possible fixes:
 * stylegan3 code updates their call to this function to give the new argument. (Not sure what would be the best way for them to support old and new versions?)
 * Revert the change in PyTorch, but this would lose the improvement from the changes.

Related issue in stylegan3 repository: https://github.com/NVlabs/stylegan3/issues/145.

**Comment by albanD** [comment]
We did change this function is an non BC-compatible way. But this function is not exposed via public API. The stylegan repo uses the internal `torch._C._jit_get_operation` to be able to access it.
I feel like the best would be to update the stylegan code to pass in the new argument (read from `ctx. needs_input_grad`).

**Comment by albanD** [comment]
> the mask should be a bool tensor with two values

The mask should be a tuple of booleans actually (as you propose in the example below). No Tensor involved here.

>  would use (ctx.needs_input_grad[1], ctx.needs_input_grad[2]) as the mask

That sounds correct yes.

**Comment by matthewlyf** [comment]
Has there any been any progress resloving this error?  I've encountered this for both stylegan2 and stylegan3.

**Comment by zikpefu** [comment]
@matthewlyf from what @albanD said, I believe this is a stylegan3 issue not a pytorch issue

**Comment by matthewlyf** [comment]
@zikpefu Were you able to find a solution in the end to get stylegan3 working?

**Comment by albanD** [comment]
You simply need to pass in the new argument based on the the ctx field as mentioned above.

**Comment by jannehellsten** [comment]
FWIW, I pushed a [change to StyleGAN3](https://github.com/NVlabs/stylegan3/commit/407db86e6fe432540a22515310188288687858fa) repo that should fix this for PyTorch 1.11.

It looks like maybe the right Python API to use would be `torch.ops.aten`.. but I didn't find an API reference for it?

**Comment by albanD** [comment]
The update looks good!
Yes `torch.ops` would allow you to access that as well. But you would have gotten the same issue :/

I didn't knew people were accessing these backward ops like that. We will be more careful in the future to be BC there as well!


Closing as this should be fixed. Feel free to open a new issue if you have any other similar problem.

**Comment by RanQChi** [comment]
> It looks like maybe the right Python API to use would be `torch.ops.aten`.. but I didn't find an API reference for it?

Thank you sooooo much!!!!! it works for me!!!!

**Comment by MarianneSheridan** [comment]
Since my question is about grid_sampler_2d_backward too, so I may ask here:

I was wondering that if I change all the `align_corners=False` into `align_corners=True` in this [file](https://github.com/NVlabs/stylegan3/blob/407db86e6fe432540a22515310188288687858fa/torch_utils/ops/grid_sample_gradfix.py), would this function be right for calculating the second derivative of F.grid_sample(align_corners=True)? If not, any clue for writing a similar function for F.grid_sample(align_corners=True)?

Best
